src/components/admin/form-components/withSelector.jsx

```diff
@@ -0,0 +1,175 @@
+
+export const withSelector = Finder => {
+  class ObjectChip extends React.PureComponent {
+    onClick = () =>
+      this.props.onDelete(this.props.id, this.props.title, this.props.slug);
+
+    render() {
+      return (
+        <Chip onRequestDelete={this.onClick} style={this.props.style}>
+          {this.props.children}
+        </Chip>
+      );
+    }
+  }
+
+  ObjectChip.propTypes = {
+    onDelete: PropTypes.func.isRequired,
+    id: PropTypes.number,
+    title: PropTypes.string.isRequired,
+    slug: PropTypes.string,
+    style: PropTypes.shape({}),
+    children: PropTypes.node,
+  };
+
+  ObjectChip.defaultProps = {
+    id: null,
+    slug: null,
+    style: {},
+    children: null,
+  };
+
+  class Selector extends React.Component {
+    constructor(props) {
+      super(props);
+      this.handleDelete = this.handleDelete.bind(this);
+      this.handleClickAdd = this.handleClickAdd.bind(this);
+    }
+
+    componentDidUpdate(prevProps) {
+      if (!this.props.disabled && this.props.elements !== prevProps.elements) {
+        this.props.onChange();
+      }
+    }
+
+    handleClickAdd = async object => {
+      // disable this if saving
+      if (this.props.disabled) return;
+
+      const { id, title, slug } = object;
+      const alreadyAdded =
+        this.props.elements.find(
+          baseObject =>
+            !('slug' in baseObject)
+              ? this.props.slugify(baseObject.name) === slug
+              : baseObject.slug === slug,
+        ) !== undefined;
+
+      if (alreadyAdded) {
+        await this.props.displayAlert(
+          `The ${this.props.mode} ${title} has already been added.`,
+        );
+        return;
+      }
+      const newObjects = update(this.props.elements, {
+        $push: [{ id, name: title, slug }],
+      });
+      this.props.onUpdate(newObjects);
+    };
+
+    handleDelete(id, title, slug) {
+      // disabled this if saving
+      if (this.props.disabled) return;
+
+      const objectSlug = !slug ? this.props.slugify(title) : slug;
+
+      const index = this.props.elements.findIndex(
+        baseObject =>
+          !('slug' in baseObject)
+            ? this.props.slugify(baseObject.name) === objectSlug
+            : baseObject.slug === objectSlug,
+      );
+      if (index === -1) {
+        throw new Error(
+          'You tried to delete an object not currently selected. ' +
+            "This shouldn't happen, please let the developers know that it did.",
+        );
+      }
+      const newObjects = update(this.props.elements, { $splice: [[index, 1]] });
+      this.props.onUpdate(newObjects);
+    }
+
+    render() {
+      const styles = {
+        wrapper: {
+          display: 'flex',
+          flexWrap: 'wrap',
+        },
+      };
+
+      const objectChips =
+        this.props.elements.length > 0
+          ? _.map(this.props.elements, object => (
+              <ObjectChip
+                key={object.id !== null ? object.id : object.name}
```

Comment:
I thought I already made a comment here but I guess I didn't press submit? Right now `id` isn't a required prop so it could also be undefined, either you should also check for it being undefined, or you should update the proptypes so that it fits.

One way is just making it required (which wouldn't allow `null`) and then making the "new object" id be -1 for example. But I'm okay with anything as long as the proptypes are correct.

Reply:
I updated it to use lodash _.get function. Ideally, it would be set to null by default through defaultProps, but I haven't been able to find documentation on defaultProp values when using an array of shapes.

Reply:
Yeah I don't think it's possible I'm afraid :(. Which is kind of ugly! And I definitely don't like splitting up setting defaults into two different places, maybe I should just disable that linting rule? But anyway, I guess using `_.get` will get you the correct behaviour if it's undefined right?

Reply:
I think it is fine to have the linting rule there. And yes, _.get defaults to object.name.

Reply:
I'm afraid not, I checked it
![screenshot from 2018-04-14 22-11-54](https://user-images.githubusercontent.com/18149242/38771255-eef2891c-4030-11e8-8373-67400be83847.png)

As requested below